### DISCUSSION TITLE
Custom discounts

### DISCUSSION THREAD

**Discussion by Deni-Iconis** [saleor/saleor#7592] [Q&A] [answered]
Hi,

(1) Is there a way to apply a custom discount(one that's neither 'sale' nor 'voucher') to the checkout via plugins?

(2) Do you plan to implement a way to limit Voucher to one specific user?

Thank you!

**Comment by TomSteer** [comment]
We've also got a similar requirement to apply a custom discount to checkout line item but from what I can tell the API only allows a custom discount to be applied to an order and not a checkout?
(https://docs.saleor.io/docs/3.0/developer/api-reference/mutations/order-line-discount-update)

**Comment by korycins** [comment]
@Deni-Iconis 
1) Plugins have interface which is responsible for calculating checkout's amounts (like `calculate_checkout_total` or `calculate_checkout_line_total`).  You could try to use it to build own custom plugin.
2) There is no explicit way to limit voucher for user, but you can create new voucher, limit usage to one, or/and limit voucher to one use per customer.

@TomSteer - We added the possibility to add discounts to Order and OrderLine (Saleor3.0). In the future, we're planning to extend this functionality to checkout also, but I don't have any specific date.

**Comment by Deni-Iconis** [comment]
@korycins 

1. Yes its possible, but it adds complexity in tracking if and when that discount is applied if you have some extra business logic to apply to it. 

1.1. what happens when you apply your discount and then add another voucher? It looks to me like it currently overwrites previously applyed discount?
It would be much easier to handle such cases if there would be an abstract discount object with ie. "is_applicable_to_checkout(checkout_info) -> bool" method that one can implement to apply some custom business logic to that discount, and a way to have multiple of such discounts applied to single checkout. 

2. I understand that and I'm using it that way, but it still is'nt as secure as limiting it to specific user if we want it to be human readable. Consider the case when customer gets code: "Maciej_K_100_EUR" vs "123e4567-e89b-12d3-a456-426614174000". If former is limited to one specific user it could be used even in a hashtag on social media(its a bit extreme example) without worrying about potential misuse by someone else.

**Comment by korycins** [comment, marked as answer]
Hey @Deni-Iconis 
1. I agree but I don't see any other possible solution for now. 
1.1 Saleor allows assigning only one voucher, at least for now.
2. Didn't know that you know about it, this is why I mentioned it :)  So currently we don't have implemented voucher per customer. AFAIK we don't have any plans to extend it for now.

**Reply by sfehlandt** [reply]
Hi @korycins 

About 1), do you have an estimate if/when the option to apply discounts to Checkouts will be ready?
For instance, is it planned for 3.0 or 3.1?

Cheers and thanks for this awesome project!
